### PR TITLE
4: Add in base definitions for first batch of FRBC messages.

### PR DESCRIPTION
**Pull request by TNO-SlaFleur**

### CODE REVIEW

**Comment by TNO-SlaFleur** [comment]
Please merge after merging #2

**Review by victorgarcia98 — COMMENTED** [review]
Thanks for starting to implement the different messages!

As a general comment, I'm not sure we should create a submodule for every class. It depends on how large we expect the classes to be. Maybe, we should keep an structures similar to this, open to alternatives:
```
common/
    __init__.py
    messages.py
    schemas.py
frbc
    __init__.py
    messages.py
    schemas.py
...
```

File: src/s2wsjson/frbc/frbc_operation_mode_element.py
[the remark this answers is not in the document]
Comment (TNO-SlaFleur) [reply src/s2wsjson/frbc/frbc_operation_mode_element.py:1]: Split to #9 !

File: src/s2wsjson/common/timer.py
[the remark this answers is not in the document]
Comment (TNO-SlaFleur) [reply src/s2wsjson/common/timer.py:1]: Refactored: https://github.com/flexiblepower/s2-ws-json-python/pull/5/commits/3965355d4853cdd035c85135572e5fedc6e2154e

**Comment by TNO-SlaFleur** [comment]
> Thanks for starting to implement the different messages!
> 
> As a general comment, I'm not sure we should create a submodule for every class. It depends on how large we expect the classes to be. Maybe, we should keep an structures similar to this, open to alternatives:
> 
> ```
> common/
>     __init__.py
>     messages.py
>     schemas.py
> frbc
>     __init__.py
>     messages.py
>     schemas.py
> ...
> ```

I thought about this as well if it would be worth it to limit the number of files. However, we expect quite a few classes so adding it to 1 or 2 files would result in large files. Navigating through these files might become cumbersome. Also, as we are adding unit tests, I do expect quite a few unit tests per class. For instance Timer currently already has 5 unit tests and is 77 lines long. Aggregating the classes into a single file would also move all the unit tests to a single file (although in separate 'class (TestCase)' structures) which may become difficult to navigate.

Right now I have a preference to split them out. For imports, I am adding it to common/__init__.py so at least the user does not need a large amount of imports. Curious to your thoughts!

**Comment by victorgarcia98** [comment]
> > Thanks for starting to implement the different messages!
> > As a general comment, I'm not sure we should create a submodule for every class. It depends on how large we expect the classes to be. Maybe, we should keep an structures similar to this, open to alternatives:
> > ```
> > common/
> >     __init__.py
> >     messages.py
> >     schemas.py
> > frbc
> >     __init__.py
> >     messages.py
> >     schemas.py
> > ...
> > ```
> 
> I thought about this as well if it would be worth it to limit the number of files. However, we expect quite a few classes so adding it to 1 or 2 files would result in large files. Navigating through these files might become cumbersome. Also, as we are adding unit tests, I do expect quite a few unit tests per class. For instance Timer currently already has 5 unit tests and is 77 lines long. Aggregating the classes into a single file would also move all the unit tests to a single file (although in separate 'class (TestCase)' structures) which may become difficult to navigate.
> 
> Right now I have a preference to split them out. For imports, I am adding it to common/**init**.py so at least the user does not need a large amount of imports. Curious to your thoughts!

Sounds like a good compromise!

By the way, feel free to send a re-review request when the PR is ready! :smile:

File: src/s2wsjson/common/transition.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+class Transition(GenTransition, ValidateValuesMixin['Transition']):
+    class Config(GenTransition.Config):
+        validate_assignment = True
+
+    id: uuid.UUID = GenTransition.__fields__['id'].field_info  # type: ignore[assignment]
+    from_: uuid.UUID = GenTransition.__fields__['from_'].field_info  # type: ignore[assignment]
+    to: uuid.UUID = GenTransition.__fields__['to'].field_info  # type: ignore[assignment]
+    start_timers: List[uuid.UUID] = GenTransition.__fields__['start_timers'].field_info  # type: ignore[assignment]
+    blocking_timers: List[uuid.UUID] = GenTransition.__fields__['blocking_timers'].field_info  # type: ignore[assignment]
+    transition_duration: Optional[Duration] = GenTransition.__fields__['transition_duration'].field_info  # type: ignore[assignment]
```
Comment (victorgarcia98) [inline src/s2wsjson/common/transition.py:19]: Shouldn't `transition_duration be of type (new) `Duration`? From this, I understand that the field would be of the old (generated)  type.

File: tests/unit/common/transition_test.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          "abnormal_condition_only": true}
+        '''
+
+        # Act / Assert
+        with self.assertRaises(S2ValidationError):
+            Transition.from_json(json_str)
+
+    def test__to_json__happy_path(self):
+        # Arrange
+        # TODO We have to resort to using a dict as we HAVE to pass the 'from' key which is a Python reserved keyword.
```
Comment (victorgarcia98) [inline tests/unit/common/transition_test.py:91]: Nice! Thanks for doing the research!

**Review by victorgarcia98 — APPROVED** [review]
Looks pretty good to me! :smile: 

This project is getting really cool!

I see there are some test files with commented tests. I don't mind going ahead and merge this and retake them in a future PR.

File: src/s2wsjson/frbc/frbc_actuator_description.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+from s2wsjson.validate_values_mixin import ValidateValuesMixin, catch_and_convert_exceptions
+
+
+@catch_and_convert_exceptions
+class FRBCActuatorDescription(GenFRBCActuatorDescription, ValidateValuesMixin['FRBCActuatorDescription']):
+    class Config(GenFRBCActuatorDescription.Config):
+        validate_assignment = True
+
+    id: uuid.UUID = GenFRBCActuatorDescription.__fields__['id'].field_info  # type: ignore[assignment]
+    operation_modes: List[FRBCOperationMode] = GenFRBCActuatorDescription.__fields__['operation_modes'].field_info  # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (TNO-SlaFleur) [reply src/s2wsjson/frbc/frbc_actuator_description.py:19]: Indeed! It keeps any of the description, alias, number restrictions or other pydantic field settings while overriding the name.

**Pull request merged**